// broker-simulator-rnymphaea/backend/dist/brokers/brokers.controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrokersController = void 0;
const common_1 = require("@nestjs/common");
const data_service_1 = require("../data/data.service");
let BrokersController = class BrokersController {
    constructor(dataService) {
        this.dataService = dataService;
    }
    getBrokers() {
        return this.dataService.getBrokers();
    }
    createBroker(broker) {
        const brokers = this.dataService.getBrokers();
        const newBroker = Object.assign(Object.assign({}, broker), { id: brokers.length ? Math.max(...brokers.map(b => b.id)) + 1 : 1 });
        brokers.push(newBroker);
        this.dataService.saveBrokers(brokers);
        return newBroker;
    }
    updateBroker(id, updateData) {
        const brokers = this.dataService.getBrokers();
        const index = brokers.findIndex(b => b.id === parseInt(id));
        if (index === -1) {
            throw new Error('Broker not found');
        }
        brokers[index] = Object.assign(Object.assign({}, brokers[index]), updateData);
        this.dataService.saveBrokers(brokers);
        return brokers[index];
    }
    deleteBroker(id) {
        const brokers = this.dataService.getBrokers();
        const filteredBrokers = brokers.filter(b => b.id !== parseInt(id));
        this.dataService.saveBrokers(filteredBrokers);
        return { success: true };
    }
};
__decorate([
    (0, common_1.Get)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Array)
], BrokersController.prototype, "getBrokers", null);
__decorate([
    (0, common_1.Post)(),
    __param(0, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], BrokersController.prototype, "createBroker", null);
__decorate([
    (0, common_1.Put)(':id'),
    __param(0, (0, common_1.Param)('id')),
    __param(1, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", Object)
], BrokersController.prototype, "updateBroker", null);
__decorate([
    (0, common_1.Delete)(':id'),
    __param(0, (0, common_1.Param)('id')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Object)
], BrokersController.prototype, "deleteBroker", null);
BrokersController = __decorate([
    (0, common_1.Controller)('brokers'),
    __metadata("design:paramtypes", [data_service_1.DataService])
], BrokersController);
exports.BrokersController = BrokersController;
