// broker-simulator-rnymphaea/backend/dist/simulation/simulation.controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimulationController = void 0;
const common_1 = require("@nestjs/common");
const data_service_1 = require("../data/data.service");
const simulation_gateway_1 = require("./simulation.gateway");
let SimulationController = class SimulationController {
    constructor(dataService, simulationGateway) {
        this.dataService = dataService;
        this.simulationGateway = simulationGateway;
    }
    getSettings() {
        return this.dataService.getSettings();
    }
    updateSettings(settings) {
        const currentSettings = this.dataService.getSettings();
        const updatedSettings = Object.assign(Object.assign({}, currentSettings), settings);
        this.dataService.saveSettings(updatedSettings);
        return updatedSettings;
    }
    startSimulation() {
        const settings = this.dataService.getSettings();
        settings.isRunning = true;
        this.dataService.saveSettings(settings);
        this.simulationGateway.startSimulation();
        return { success: true };
    }
    stopSimulation() {
        const settings = this.dataService.getSettings();
        settings.isRunning = false;
        this.dataService.saveSettings(settings);
        this.simulationGateway.stopSimulation();
        return { success: true };
    }
};
__decorate([
    (0, common_1.Get)('settings'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Object)
], SimulationController.prototype, "getSettings", null);
__decorate([
    (0, common_1.Put)('settings'),
    __param(0, (0, common_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], SimulationController.prototype, "updateSettings", null);
__decorate([
    (0, common_1.Post)('start'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Object)
], SimulationController.prototype, "startSimulation", null);
__decorate([
    (0, common_1.Post)('stop'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Object)
], SimulationController.prototype, "stopSimulation", null);
SimulationController = __decorate([
    (0, common_1.Controller)('simulation'),
    __metadata("design:paramtypes", [data_service_1.DataService,
        simulation_gateway_1.SimulationGateway])
], SimulationController);
exports.SimulationController = SimulationController;
