// broker-simulator-rnymphaea/backend/dist/simulation/simulation.gateway.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimulationGateway = void 0;
const websockets_1 = require("@nestjs/websockets");
const socket_io_1 = require("socket.io");
const data_service_1 = require("../data/data.service");
let SimulationGateway = class SimulationGateway {
    constructor(dataService) {
        this.dataService = dataService;
    }
    afterInit(server) {
        console.log('WebSocket server initialized');
    }
    handleConnection(client) {
        console.log(`Client connected: ${client.id}`);
        this.sendCurrentData();
    }
    handleDisconnect(client) {
        console.log(`Client disconnected: ${client.id}`);
    }
    startSimulation() {
        const settings = this.dataService.getSettings();
        const stocks = this.dataService.getStocks();
        let startIndex = 0;
        if (settings.startDate && stocks.length > 0 && stocks[0].historicalData) {
            const startDateIndex = stocks[0].historicalData.findIndex(data => data.date === settings.startDate);
            if (startDateIndex !== -1) {
                startIndex = startDateIndex;
            }
        }
        settings.currentDateIndex = startIndex;
        settings.isRunning = true;
        this.dataService.saveSettings(settings);
        console.log(`Simulation started from date: ${settings.startDate}, index: ${startIndex}`);
        this.simulationInterval = setInterval(() => {
            this.nextDay();
        }, settings.speed * 1000);
    }
    stopSimulation() {
        if (this.simulationInterval) {
            clearInterval(this.simulationInterval);
        }
        const settings = this.dataService.getSettings();
        settings.isRunning = false;
        this.dataService.saveSettings(settings);
    }
    nextDay() {
        var _a;
        const settings = this.dataService.getSettings();
        const stocks = this.dataService.getStocks();
        if (settings.currentDateIndex >= stocks[0].historicalData.length - 1) {
            this.stopSimulation();
            console.log('Simulation finished - reached end of historical data');
            return;
        }
        settings.currentDateIndex += 1;
        this.dataService.saveSettings(settings);
        const currentStocksData = stocks.map(stock => {
            const historicalData = stock.historicalData[settings.currentDateIndex];
            return {
                id: stock.id,
                symbol: stock.symbol,
                name: stock.name,
                currentPrice: historicalData ? historicalData.open : 0,
                date: historicalData ? historicalData.date : 'N/A',
            };
        });
        this.server.emit('stockUpdate', currentStocksData);
        console.log(`Day ${settings.currentDateIndex}: ${(_a = currentStocksData[0]) === null || _a === void 0 ? void 0 : _a.date}`);
    }
    sendCurrentData() {
        const settings = this.dataService.getSettings();
        const stocks = this.dataService.getStocks();
        const currentStocksData = stocks.map(stock => {
            const historicalData = stock.historicalData[settings.currentDateIndex];
            return {
                id: stock.id,
                symbol: stock.symbol,
                name: stock.name,
                currentPrice: historicalData ? historicalData.open : 0,
                date: historicalData ? historicalData.date : 'N/A',
            };
        });
        this.server.emit('stockUpdate', currentStocksData);
    }
};
__decorate([
    (0, websockets_1.WebSocketServer)(),
    __metadata("design:type", socket_io_1.Server)
], SimulationGateway.prototype, "server", void 0);
SimulationGateway = __decorate([
    (0, websockets_1.WebSocketGateway)({ cors: true }),
    __metadata("design:paramtypes", [data_service_1.DataService])
], SimulationGateway);
exports.SimulationGateway = SimulationGateway;
